Add s3-service.md to document the S3 service inside Safe Haven Services

howff opened the pull request:
# EIDF Documentation Pull Request

## Description

Add s3-service.md to document the S3 service inside Safe Haven Services

Fixes #237 237

## Type of change

Please delete options that are not relevant.

- [ ] Incorrect Documentation
- [ ] Incomplete Documentation
- [X] Missing Documentation
- [ ] Formatting
- [ ] New Documentation

## What has to be reviewed

s3-service.md

## Checklist

- [ ] Documentation follows the project style guidelines
- [ ] Ensure Contact details contain Service Emails and Numbers
- [ ] Self-review of documentation using mkdocs on local system
- [ ] Spellcheck has been performed
- [ ] Pre-commit has been run and passed

howff committed:
Added s3-service.md

2bPro commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,145 @@
+# S3 Service
+
+There is no general-purpose S3 service within Safe Haven Services, unlike the [EIDF S3 service](../services/s3/).
+
+However there is a S3 service in SHS with the following caveats:
+* it is only available to the Scottish National Safe Haven, other tenants by arrangement
+* it is a read-only service, as a way of providing access to large collections of files
+* it is not a storage solution for users wanting to create their own files
+
+## Access arrangements
+
+Access to buckets is via keys (Access Key and Secret Access Key) provided to the user by the Research Coordinator.
```
Suggested rewording: Some Safe Havens may provide you with access to data via S3. If this applies to your project, your Research Coordinator will provide you with an access key. This documentation will guide you through how to get access to your data from a terminal as well as programmatically via R and Python.

!!! important Files in S3 buckets are read-only. If you need to transform or make changes to any files, you will need to download them to your project space. If you download files, please be mindful of disk space by only downloading what is necessary and deleting them as soon as no longer needed.

2bPro commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,145 @@
+
+Access to buckets is via keys (Access Key and Secret Access Key) provided to the user by the Research Coordinator.
+
+## How to use the service
```
## Environment setup

2bPro commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,145 @@
+| AWS_SECRET_ACCESS_KEY | as provided by RC    |
+| http_proxy            | no value             |
+
+```
+export AWS_DEFAULT_REGION=us-east-1
+export AWS_ENDPOINT_URL=http://nsh-fs02:7070
+export AWS_ACCESS_KEY_ID=put_your_key_here
+export AWS_SECRET_ACCESS_KEY=put_your_secret_here
+export http_proxy=
+```
````
Is there a better, more permanent solution within the constraints of the user? For example, is it possible to run `aws configure --profile minio` and add the endpoint to ~/.aws/config?

2bPro replied:
To discuss for future improvement. The docs can go ahead as they are just now. Can we say between the table and the code "Create an env file with the following variables:" and after "Apply the configuration by running `source s3.env`" or similar?

2bPro commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,145 @@
+
+Check it's installed, run `aws` and it will show some help. If you get `command not found` then run it as `~/.local/bin/aws`.
+
+The general syntax is `aws s3 cp s3://bucket/filename localfilename`, for example:
+```
+aws s3 cp  s3://extraction_5_CT/123/456/789.dcm  copy_of_789.dcm
+```
````
To download a file, run:

```console
aws s3 cp s3://bucket/filename localfilename
```

2bPro commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,145 @@
+* If you need to save into a file temporarily (e.g. whilst converting to NIFTI) then save into a RAM disk in `/run/user/$(id -u)/` but delete it straight after use to recover then memory.
+* If it's too large for RAM then save into a file on the system disk, not in your home directory, in `/tmp/$(id)/` but check the disk has space first (using `df -h /tmp/`) and delete it straight after use to recover the disk space.
+
+## Python Usage
```
### Python

2bPro commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,145 @@
+
+## Python Usage
+
+### Setup
```
First, install `boto3`:

Commit:
Update s3-service.md

2bPro commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,145 @@
+```
+
+Check it's installed, run `aws` and it will show some help. If you get `command not found` then run it as `~/.local/bin/aws`.
+
````
Was this not added due to the number of files in the bucket? The SDP seems to indicate that the user would know the contents of the bucket from a CSV file. This is not mentioned here, and the copy command below assumes the user somehow knows exactly what is in the bucket and the path. It may also be good to give a link to [the documentation](https://awscli.amazonaws.com/v2/documentation/api/2.9.6/reference/s3/ls.html#:~:text=%2D%2Dpage%2Dsize%20(integer),sizes%20in%20human%20readable%20format.).

Commit:
Update s3-service.md, especially for R examples

howff commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,145 @@
+If this command fails it might be due to a proxy configuration in your environment. To temporarily turn off the proxy in the current window use this first:
+```
+export http_proxy=
+```
````
I don't know if it's technically possible for the web proxy to be configured to pass traffic onto nsh-fs02. Maybe a question for Barry or similar. If so it would reduce confusion, but on the other hand I don't think it's a good idea to put the web proxy between the client and the S3 server because all it does is cause additional unnecessary load on the proxy and slow everything down; there's no benefit from authentication either.

2bPro replied:
Had a chat with Barry about this, he recommends the use of the NO_PROXY variable set in bashrc instead, so this would look like `export NO_PROXY="$NO_PROXY,nsh-fs02:7070"`. I also asked Susan, and users are never told to go and edit their bashrcs, which we may want to avoid. The other option is for systems to add this to their bashrcs (either to all or on-demand).

howff replied:
Last time I tested this the NO_PROXY variable was ignored!

rkm replied:
NO_PROXY will work here, if properly configured:

## Routes to the proxy

```console
rmacleod@nsh-rc-desktop01:~$ NO_PROXY=''; curl -LI nsh-fs02:7070
HTTP/1.1 503 Service Unavailable
Server: squid
...
```
## Routes directly to nsh-fs02

```console
rmacleod@nsh-rc-desktop01:~$ NO_PROXY=nsh-fs02; curl -LI nsh-fs02:7070
HTTP/1.1 400 Bad Request
Server: VERSITYGW
...
```

A more generalised solution would be to set `NO_PROXY=localhost,127.0.0.1,.nsh.loc` and then to fully-qualify the server as `nsh-fs02.nsh.loc`.

howff replied:
NO_PROXY doesn't work; R ignores it and Python ignores it. awscli obeys it. Python does obey no_proxy, R ignores it.

Commit:
Update s3-service.md, especially for Python examples

Commit:
Add safe-haven-services/s3-service.md to the navigation page via mkdocs.yml

Commit:
Update s3-service.md, env vars specifically for R aws.s3

Commit:
Update s3-service.md, regarding http_proxy

2bPro commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,214 @@
+save_object( my_object_path, file = "downloaded.dcm", bucket = my_bucket, base_url=my_endpoint_host, region="", use_https=FALSE, key = my_access_key, secret = my_secret_key )
+```
+
+Note! You need to have the region set in the environment variable and pass `region=""` to the functions, otherwise you get a `cannot resolve host` error. The `base_url` is just the host and port, no `http://` prefix, and `use_https` is false.
````
Note formatting is inconsistent within this page as well as compared with the rest of the docs. For important pieces of text, [this](https://docs.eidf.ac.uk/services/gpuservice/policies/#kubernetes-job-names) is probably the most visible.

2bPro commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,211 @@
+To access files you need to know:
+* An "endpoint URL" (for the NSH this is `http://nsh-fs02:7070`)
+* The "Region" of the service (for the NSH this is `us-east-1`)
+* A "bucket" name (similar to a folder name)
+* An "access key" (similar to a username)
+* A "private key" (similar to a password)
```
When defining the variables, a user needs:
AWS_ACCESS_KEY_ID=bucketname
AWS_SECRET_ACCESS_KEY=password

howff replied:
The bucket name is like the directory name, the access key is like the username and the secret key is like the password, I'm not sure where you're getting the information that the access key is the same as the bucket name.

2bPro replied:
As specified in my previous comment, I'm getting it from the documentation and the file received from the RC. Maybe an example would help clarify where I'm coming from.

I have a user `bprodan` in study `2024-0000` and was told by the RC that I will be given access to a bucket called `dummydata`. According to the user documentation, which follows the standard minio convention, the configuration should be:

| variable | value |
| -------- | ----- |
| BUCKET_NAME | `dummydata` |
| ACCESS_KEY_ID | `bprodan` |
| SECRET_ACCESS_KEY | `1234` |

According to the RC documentation, keys are created at study level, so this should be more like:
| variable | value |
| -------- | ----- |
| BUCKET_NAME | `dummydata` |
| ACCESS_KEY_ID | `2024-0000` |
| SECRET_ACCESS_KEY | `1234` |

But instead, in the file received from the RC, I have:
| variable | value |
| -------- | ----- |
| BUCKET_NAME | `dummydata` |
| ACCESS_KEY_ID | `dummydata` |
| SECRET_ACCESS_KEY | `1234` |

Currently, the user documentation assumes knowledge of how minio is configured and calling the access key ID a username just to then receive the bucket name would make a user question which value maps to that variable and if that is their username or not.

rkm replied:
>The bucket name is like the directory name, the access key is like the username and the secret key is like the password, I'm not sure where you're getting the information that the access key is the same as the bucket name.

I agree with this, so perhaps the RC documentation needs to be updated?

As an aside, we are not using MinIO.

2bPro replied:
The technology is not important in this case, this is about user guidance on how to configure access. I am not disagreeing with Andrew's statement, I'm saying the way the documentation presents this assumes knowledge about what the access key ID variable represents and is unclear about what value is associated with it.

2bPro commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,214 @@
+```py
+import boto3
+import os
+os.environ["AWS_ENDPOINT_URL"] = "http://nsh-fs02:7070"
+os.environ["AWS_DEFAULT_REGION"] = "us-east-1"
+os.environ["AWS_ACCESS_KEY_ID"] = "dummydata"
+os.environ["AWS_SECRET_ACCESS_KEY"] = "put_your_secret_here"
+os.environ["http_proxy"] = ""
````
The user was already told to set the environment variables. Does boto automatically pick these up or is something like [this ](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#passing-credentials-as-parameters) needed?

kavousan reviewed (commented):
If the answer to the pip install comment is that it's not a problem, this is an approval.

2bPro commented on docs/safe-haven-services/s3-service.md:
Yes, anyone can run pip install as part of access to PyPi via the web proxy.

Commit:
Simplified the S3 service docs

howff committed:
s3 docs - Add GUI example for viewing DICOM files from S3

Added instructions for using a GUI to view DICOM files from the S3 service, including container setup and script creation.

howff committed:
Merge branch 'main' into 237-documentation-the-prototype-s3-service-in-the-nsh

kavousan commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,246 @@
+All of the above will be supplied to you by your Research Coordinator.
+It should go without saying that the access key details are confidential and must never be shared or allowed to be seen by others.
+Note that all file accesses are logged.
+
```
I would replace this with a drier note: " S3 access key details are confidential and must never be shared or allowed to be seen by others. All file accesses are logged."

kavousan commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,246 @@
+## Performance tips
+
+* consume the file directly in memory if possible. Saving to disk is not recommended; it wastes disk space and will take 3 times longer to do your processing. See the example code below.
+* If you need to save into a file temporarily (e.g. whilst converting to NIFTI) then save into a RAM disk in `/run/user/$(id -u)/` but delete it straight after use to recover then memory.
```
...to recover memory. (or recover the memory, not sure what applies)

kavousan commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,246 @@
+
+* consume the file directly in memory if possible. Saving to disk is not recommended; it wastes disk space and will take 3 times longer to do your processing. See the example code below.
+* If you need to save into a file temporarily (e.g. whilst converting to NIFTI) then save into a RAM disk in `/run/user/$(id -u)/` but delete it straight after use to recover then memory.
+* If it's too large for RAM then save into a file on the system disk, not in your home directory, in `/tmp/$(id)/` but check the disk has space first (using `df -h /tmp/`) and delete it straight after use to recover the disk space.
```
Is /tmp/$(id)/ only readable by the user (and maybe their group)?

kavousan commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -0,0 +1,246 @@
+## Example GUI to view DICOM files
+
+This example uses a customised program that can view and download DICOM files from the S3 service.
+It is not installed in the National Safe Haven but you can import it as a container image.
```
As per first sentence, this appears to be drafted as a Tenant-agnostic page, so replace with "If it is not installed in your Safe Haven, you can import is as a container image."

kavousan commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,246 @@
+REPOSITORY                TAG    IMAGE ID      CREATED       SIZE
+ghcr.io/howff/dcmaudit    cpu    4f994194efa8  11 days ago   2.67 GB
+```
+Now you will need a 's3' directory in your home directory so create that, or run:
````
so to create that, run the following command:

kavousan commented on docs/safe-haven-services/s3-service.md:
````diff
@@ -0,0 +1,246 @@
+```
+ces-pm-run --opt-file <(echo -v $HOME/.dcmaudit:/root/.dcmaudit -v $HOME/s3:/root/s3 --http-proxy=false) ghcr.io/howff/dcmaudit:cpu
+```
+That is difficult to type every time you need it, so we will create a script:
````
so you can create a script as follows:

kavousan reviewed (changes requested):
I made some simple suggestions and let more minor points go. If there is no time, and if the answer to the /tmp comment is that only user (and maybe their group, if IG happy with that) can read this dir, I can approve without these changes/.

howff committed:
Update wording of S3 service document

kavousan commented on docs/safe-haven-services/s3-service.md:
```diff
@@ -49,8 +49,8 @@ Many S3 tools will read environment variables so this can be a convenient way to
-* If you need to save into a file temporarily (e.g. whilst converting to NIFTI) then save into a RAM disk in `/run/user/$(id -u)/` but delete it straight after use to recover then memory.
-* If it's too large for RAM then save into a file on the system disk, not in your home directory, in `/tmp/$(id)/` but check the disk has space first (using `df -h /tmp/`) and delete it straight after use to recover the disk space.
+* If you need to save into a file temporarily (e.g. whilst converting to NIFTI) then save into a RAM disk in `/run/user/$(id -u)/` but delete it straight after use to recover the memory.
+* If it's too large for RAM then save into a file on the system disk, not in your home directory, for example in `/tmp/$(id -un)/` but check the disk has space first (using `df -h /tmp/`) and delete it straight after use to recover the disk space.
```
I spoke to Donald. The Safe Setting does not enforce permissions on /tmp; it would be at least a near-miss to put Safe Data there without correct permissions. There is also no real performance benefit between /tmp and the user account. This line should be deleted.

Donald's advice is to delete both lines; the advice should be to just work in memory.

howff committed:
Revise performance tips for S3 service

Don't suggest a private directory underneath /tmp

kavousan reviewed (approved):
Many thanks to all who contributed text and reviews.
